File: CSE 208/Graph Practice/graph_practice.cpp
/**
 * Jan25 CSE208
 * Practice Session on Graphs
 *
 *
 * Complete the following code by implementing the functions. Please be mindful of the corner cases.
 * Run the provided input file.
 * Be ready to explain the generated output to the evaluator.
 *
 *
 * The usage of any online tool and/or any GenAI is absolutely forbidden in this course
 * unless otherwise mentioned.
*/

#include <iostream>
#include <vector>
using namespace std;

class Graph {
    int V;
    vector<vector<int>>adj;
    vector<int> restricted;

public:
    Graph(int vertices){
        for(int i=0;i<vertices;i++){
            vector<int> a(vertices,0);
            adj.push_back(a);
        }
        for(int i=0;i<vertices;i++){
            restricted.push_back(0);
        }
    }

    void addEdge(int u, int v){
        if(u<0 || v < 0 || u >= adj.size() || v >= adj.size() || restricted[u] == 1 || restricted[v] == 1){
            cout << "Invalid vertex\n";
            return ;
        }
        adj[u][v]=1;
    }
    void removeEdge(int u, int v){
        if(u < 0 || v < 0 || u >= adj.size() || v >= adj.size()){
            cout << "Invalid vertex\n";
            return;
        }
        adj[u][v]=0;
    }

    int inDegree(int u){
        if(u < 0 || u >= adj.size() || restricted[u]==1){
            return -1;
        }
        int count=0;
        for(int i=0;i<adj.size();i++){
            if(adj[i][u]==1)count++;
        }
        return count;
    }
    int outDegree(int u){
        if(u < 0 || u >= adj.size() || restricted[u]==1){
            return -1;
        }
        int count=0;
        for(int i=0;i<adj[u].size();i++){
            if(adj[u][i]==1)count++;
        }
        return count;
    }


    bool removeVertex(int u){
        if(u<0 || u >= restricted.size() || restricted[u] == 1){
            return false;
        }
        for(int i=0;i<adj.size();i++){
            adj[i][u] = 0;
            adj[u][i] = 0;
        }
        restricted[u]=1;
        return true;
    }

    void printGraph(){
        for(int i=0;i<adj.size();i++){
            if(restricted[i]==1)continue;
            cout << i << " -> ";
            for(int j=0;j<adj[i].size();j++){
                if(restricted[j]!=1 && adj[i][j]==1){
                    cout<< j << " ";
                }
            }
            cout << endl;
        }
    }
    void printInNeighbors(int u){
        if(u < 0 || u >= adj.size() || restricted[u]==1){
            cout << "Node doesnt exist\n";
            return;
        }
        for(int i=0;i<adj.size();i++){
            if(adj[i][u]==1 && restricted[i]==0){
                cout << i <<" ";
            }

        }
        cout << "\n";
    }
    void printOutNeighbors(int u){
        if(u < 0 || u >= adj.size() || restricted[u]==1){
            cout << "Node doesnt exist\n";
            return;
        }
        for(int i=0;i<adj[u].size();i++){
            if(adj[u][i]==1 && restricted[i]==0){
                cout << i <<" ";
            }

        }
        cout << "\n";
    }

    // please note that once a vertex is removed, it will not be printed in the printGraph function
};

int main() {
    int initialVertices;
    cout << "Enter number of initial vertices: ";
    cin >> initialVertices;
    Graph g(initialVertices);

    while (true) {
        cout << "\n========== MENU ==========\n";
        cout << "1. Add Edge\n";
        cout << "2. Remove Edge\n";
        cout << "3. Remove Vertex\n";
        cout << "4. Print Graph\n";
        cout << "5. Find In-Degree of a Vertex\n";
        cout << "6. Find Out-Degree of a Vertex\n";
        cout << "7. Print In-Neighbors of a Vertex\n";
        cout << "8. Print Out-Neighbors of a Vertex\n";
        cout << "9. Exit\n";
        cout << "Choose an option: ";

        int choice, u, v;
        cin >> choice;

        if (choice == 9) break;

        switch (choice) {
            case 1:
                cout << "Enter source and destination vertex: ";
                cin >> u >> v;
                g.addEdge(u, v);
                break;
            case 2:
                cout << "Enter source and destination vertex to remove edge: ";
                cin >> u >> v;
                g.removeEdge(u, v);
                break;
            case 3:
                cout << "Enter vertex to remove: ";
                cin >> u;
                if (g.removeVertex(u)) {
                    cout << "Vertex " << u << " removed successfully.\n";
                } else {
                    cout << "Failed to remove vertex " << u << ".\n";
                }
                break;
            case 4:
                g.printGraph();
                break;
            case 5:
                cout << "Enter vertex to find in-degree: ";
                cin >> v;
                cout << "In-degree of " << v << " is: " << g.inDegree(v) << "\n";
                break;
            case 6:
                cout << "Enter vertex to find out-degree: ";
                cin >> v;
                cout << "Out-degree of " << v << " is: " << g.outDegree(v) << "\n";
                break;
            case 7:
                cout << "Enter vertex to find in-neighbors: ";
                cin >> u;
                g.printInNeighbors(u);
                break;
            case 8:
                cout << "Enter vertex to find out-neighbors: ";
                cin >> u;
                g.printOutNeighbors(u);
                break;
            default:
                cout << "Invalid choice. Try again.\n";
        }
    }

    cout << "Program exited.\n";
    return 0;
}

File: CSE 208/Graph Practice/GraphPractice.java
/**
 * Jan25 CSE208 
 * Practice Session on Graphs
 * 
 * 
 * Complete the following code by implementing the methods. Please be mindful of the corner cases. 
 * Run the provided input file. 
 * Be ready to explain the generated output to the evaluator.
 * 
 * 
 * The usage of any online tool and/or any GenAI is absolutely forbidden in this course 
 * unless otherwise mentioned. 
*/

import java.util.ArrayList;
import java.util.Scanner;

abstract class AbstractGraph {
    protected int V;
    protected ArrayList<ArrayList<Integer>> adj;

    public AbstractGraph(int vertices) {
        this.V = vertices;
        this.adj = new ArrayList<>(vertices);
        for (int i = 0; i < vertices; i++) {
            adj.add(new ArrayList<>());
        }
    }

    // Abstract methods
    public abstract void addEdge(int u, int v);
    public abstract void removeEdge(int u, int v);
    public abstract int inDegree(int u);
    public abstract int outDegree(int u);
    public abstract boolean removeVertex(int u);
    public abstract void printGraph();
    public abstract void printInNeighbors(int u);
    public abstract void printOutNeighbors(int u);

    // please note that once a vertex is removed, it will not be printed in the printGraph function 
}

// extend AbstractGraph into a concrete Graph class 
class Graph extends AbstractGraph {
    private ArrayList<Integer> restricted;
    public Graph(int vertices){
        super(vertices);
        restricted = new ArrayList<>();
        for(int i=0;i<vertices;i++){
            restricted.add(0);
        }
    }
    public void addEdge(int u,int v){
        if(u<0 || v<0 || u>=V || v>=V || restricted.get(u)==1 || restricted.get(v)==1)
            return;
        adj.get(u).add(v);
    }
    public void removeEdge(int u,int v){
        if(u<0 || v<0 || u>=V || v>=V)
            return;
        adj.get(u).remove(Integer.valueOf(v));
    }
    public int outDegree(int u){
        if(u < 0 || u >= V || restricted.get(u)==1)
            return -1;
        return adj.get(u).size();
    }
    public int inDegree(int u){
        if(u < 0 || u >= V || restricted.get(u)==1)
        return -1;
        int count=0;
        for(int i=0;i<V;i++){
            if(adj.get(i).contains(u))
                count++;
        }
        return count;
    }
    public boolean removeVertex(int u){
        if(u<0 || u>=V || restricted.get(u)==1)
            return false;
        restricted.set(u,1);
        for(int i=0;i<V;i++){
            adj.get(i).remove(Integer.valueOf(u));
        }
        adj.get(u).clear();
        return true;
    }
    public void printGraph(){
        for(int i=0;i<V;i++){
            if(restricted.get(i)==1)
                continue;
            System.out.print(i+" -> ");
            for(int x:adj.get(i)){
                if(restricted.get(x)==0)
                    System.out.print(x+" ");
            }
            System.out.println();
        }
    }
    public void printInNeighbors(int u){
        if(u < 0 || u >= V || restricted.get(u)==1){
            System.out.println("Node doesnt exist");
            return;
        }
        for(int i=0;i<V;i++){
            if(adj.get(i).contains(u))
                System.out.print(i+" ");
        }
        System.out.println();
    }
    public void printOutNeighbors(int u){
        if(u < 0 || u >= V || restricted.get(u)==1){
            System.out.println("Node doesnt exist");
            return;
        }
        for(int x:adj.get(u))
            System.out.print(x+" ");
        System.out.println();
    }
}


public class GraphPractice {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter number of initial vertices: ");
        int initialVertices = scanner.nextInt();
        Graph g = new Graph(initialVertices);

        while (true) {
            System.out.println("\n========== MENU ==========");
            System.out.println("1. Add Edge");
            System.out.println("2. Remove Edge");
            System.out.println("3. Remove Vertex");
            System.out.println("4. Print Graph");
            System.out.println("5. Find In-Degree of a Vertex");
            System.out.println("6. Find Out-Degree of a Vertex");
            System.out.println("7. Print In-Neighbors of a Vertex");
            System.out.println("8. Print Out-Neighbors of a Vertex");
            System.out.println("9. Exit");
            System.out.print("Choose an option: ");

            int choice = scanner.nextInt();

            if (choice == 9) break;

            switch (choice) {
                case 1:
                    System.out.print("Enter source and destination vertex: ");
                    int u1 = scanner.nextInt();
                    int v1 = scanner.nextInt();
                    g.addEdge(u1, v1);
                    break;
                case 2:
                    System.out.print("Enter source and destination vertex to remove edge: ");
                    int u2 = scanner.nextInt();
                    int v2 = scanner.nextInt();
                    g.removeEdge(u2, v2);
                    break;
                case 3:
                    System.out.print("Enter vertex to remove: ");
                    int u3 = scanner.nextInt();
                    if (g.removeVertex(u3)) {
                        System.out.println("Vertex " + u3 + " removed successfully.");
                    } else {
                        System.out.println("Failed to remove vertex " + u3 + ".");
                    }
                    break;
                case 4:
                    g.printGraph();
                    break;
                case 5:
                    System.out.print("Enter vertex to find in-degree: ");
                    int v5 = scanner.nextInt();
                    System.out.println("In-degree of " + v5 + " is: " + g.inDegree(v5));
                    break;
                case 6:
                    System.out.print("Enter vertex to find out-degree: ");
                    int v6 = scanner.nextInt();
                    System.out.println("Out-degree of " + v6 + " is: " + g.outDegree(v6));
                    break;
                case 7:
                    System.out.print("Enter vertex to find in-neighbors: ");
                    int u7 = scanner.nextInt();
                    g.printInNeighbors(u7);
                    break;
                case 8:
                    System.out.print("Enter vertex to find out-neighbors: ");
                    int u8 = scanner.nextInt();
                    g.printOutNeighbors(u8);
                    break;
                default:
                    System.out.println("Invalid choice. Try again.");
            }
        }

        System.out.println("Program exited.");
        scanner.close();
    }
}